Text splitting class

zawadzki-dawid opened the pull request:
Ciao fratelli miei, I commited some changes based on your advices. I'm not sure if I understood you well but I hope code will meet your expectations this time :love_you_gesture:

zawadzki-dawid committed:
Add Header, Source and CMakeLists files corresponding to TextSplitter class

zawadzki-dawid committed:
Update CMakeLists files to detect TextSplitter

zawadzki-dawid committed:
Implement TextSplitter class Header and Body files

zawadzki-dawid committed:
Add comment to splitText method in TextSplitter source file

szymon-filipiak commented on header/Utility/TextSplitter.h:
```diff
@@ -0,0 +1,25 @@
+class TextSplitter {
+
+private:
+    std::vector<std::tuple<long long int, long long int>> ranges;
```
So actually the types stored here do not require to be `long long int`. It's better to use type `size_t` as it's type for storing index values. It is guaranted to have range as big as the largest memory block that you can allocate on the device. It's an unsigned type.

szymon-filipiak commented on src/Utility/TextSplitter.cpp:
```diff
@@ -0,0 +1,29 @@
+//
+// Created by dawid on 23.10.2020.
+//
```
🚨

zawadzki-dawid replied:
Ay, oversight :/

szymon-filipiak commented on header/Utility/TextSplitter.h:
```diff
@@ -0,0 +1,25 @@
+
+    int rangeWidth(long long int textLength);
+public:
+    std::vector<std::tuple<long long int, long long int>>& splitText(long long int textLength, int patternLength);
```
I think we can create the vector in function and then return it by value(with move semantic underneath).

zawadzki-dawid replied:
Yep, it's one of the possibilities, why do you think it'd better approach? Is it about reusability?

szymon-filipiak commented on src/Utility/TextSplitter.cpp:
```diff
@@ -0,0 +1,29 @@
+        }
+        else if((i + patternLength - 1) < textLength && ((i + width - 1 + patternLength - 1) >= textLength)) {
+            ranges.push_back(std::make_tuple(i, textLength - patternLength));
+        } else if((i + patternLength - 1) >= textLength) {
```
I don't like this part with putting the invalid pair. We can make this algorithm produce only valid ranges and throw an error if something is really wrong. I have an idea on how to do this. We need to operate on integers. With a integer division operator you can calculate the len of the chain and then use the reminder from this operation to see how many chains must be 1 window longer

szymon-filipiak commented on src/Utility/TextSplitter.cpp:
```diff
@@ -0,0 +1,29 @@
+#include "TextSplitter.h"
+
+int TextSplitter::rangeWidth(long long int textLength) {
+    return ceil((double) textLength / (double) numOfThreads);
```
C style casts are not a recommended way to cast expressions. See `std::static_cast`

zawadzki-dawid replied:
Okey, copy that :+1:

szymon-filipiak commented on src/Utility/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+add_library(utility STATIC
```
Nice use of cmakelists library convention.

zawadzki-dawid replied:
<3

szymon-filipiak commented on src/Utility/TextSplitter.cpp:
```diff
@@ -0,0 +1,29 @@
+// It's hard cut the text into substrings of equal size without getting out of range sometimes,
+// so when we get out of range a <-1, -1> is pushed back to indicate that this tuple is wrong and thread should not process it
+
+std::vector<std::tuple<long long int, long long int>>& TextSplitter::splitText(long long int textLength, int patternLength) {
```
Also we can use std::pair here.

zawadzki-dawid replied:
We can, or we should? Is there a big difference? Teach me c++ master :D

szymon-filipiak replied:
We can. I think `std::pair` is cleaner. The difference in performance is non-existant. `std::tuple` can be used to store more than 2 values. Std::pair is  always well... a pair :P It provides easier access to the variables by memeber functions `.first()` and `.second()`.  To access the values stored in `std::tuple` you must use a friend function `get<>()`. Pair implementation is based on tuple

zawadzki-dawid replied:
Okey, I think that std::pair looks more neat as well, I'll change that

zawadzki-dawid committed:
Remove auto generated comment

zawadzki-dawid committed:
Replace long long int with size_t in rangeWidth method

zawadzki-dawid committed:
Remove range property inTextSplitter class due to change in approach to a problem

zawadzki-dawid committed:
Remake TextSplitter class for better performance

1. Add numOfWindows method to calculate number of windows to examine in text
2. Add lenOfChain method to calculate number of chains for one thread and number
   of additional windows
3. Refactor code of splitText method to make it more efficient and easier to handle by removing unneccesary if statments

zawadzki-dawid committed:
Remove comment that is no longer appropraite

WiktorPieklik commented on header/Utility/TextSplitter.h:
```diff
@@ -0,0 +1,25 @@
+private:
+    static const int numOfThreads = 32;
+
+    int numOfWindows(long long int textLength, int patternLength);
```
Since text to be searched can be of `long long int` length, it is possible to search for pattern of `long long int` length. Also it is extremely possible that having `long long int textLength` will cause generating `long long int numOfWindows`.

szymon-filipiak replied:
This is also case where we should use `size_t`.

zawadzki-dawid replied:
Yeah, you're right, oversight. Thanks for being alert!

WiktorPieklik commented on src/Utility/TextSplitter.cpp:
```diff
@@ -0,0 +1,33 @@
+}
+
+std::pair<int, int> TextSplitter::lenOfChain(int windows) {
+    int rangeWidth = windows / numOfThreads;
```
`window` is represented by `std::tuple<size_t, size_t>` so in here please use something like `numOfWindows` just like you do it with `numOfThreads`. Current naming convention is misleading. Please do it everywhere in code where `windows` stands for `number of windows`.

WiktorPieklik commented on src/Utility/TextSplitter.cpp:
```diff
@@ -0,0 +1,33 @@
+std::vector<std::tuple<size_t, size_t>> TextSplitter::splitText(long long int textLength, int patternLength) {
+    std::vector<std::tuple<size_t, size_t>> ranges;
+    int windows = numOfWindows(textLength, patternLength);
+    std::pair<int, int> chains = lenOfChain(windows);
```
Just like with `windows`, change `chains` to something like `chainsLengths`. This is misleading that `lenOfChain` method returns `chains`.

WiktorPieklik commented on src/Utility/TextSplitter.cpp:
```diff
@@ -0,0 +1,33 @@
+    return textLength - patternLength + 1;
+}
+
+std::pair<int, int> TextSplitter::lenOfChain(int windows) {
```
`lenOfChain()` returns `std::pair<int, int>` so please rename it to something like `lenOfChains()`.

szymon-filipiak replied:
Also `size_t` here.

WiktorPieklik commented on header/Utility/TextSplitter.h:
```diff
@@ -0,0 +1,25 @@
+//
+// Created by dawid on 23.10.2020.
+//
```
🚨  🚨  🚨

WiktorPieklik commented on src/CMakeLists.txt:
```diff
@@ -1,10 +1,10 @@
 add_subdirectory(Hash)
 add_subdirectory(Text)
-
+add_subdirectory(Utility)
```
I kindly ask you to consider renaming directory `Utility` to `Utils`. In my opinion it would be more appropriate. Let me know what you think 😄

szymon-filipiak replied:
It's either way for me. Utility is good and Utils is also good

zawadzki-dawid replied:
`Utility` is staying, I think it suits better current naming convention :100:

WiktorPieklik replied:
I was out-voted 😆

WiktorPieklik commented on src/Utility/TextSplitter.cpp:
```diff
@@ -0,0 +1,33 @@
+    std::vector<std::tuple<size_t, size_t>> ranges;
+    int windows = numOfWindows(textLength, patternLength);
+    std::pair<int, int> chains = lenOfChain(windows);
+    int currenPosition = 0;
```
You're missing `t` in here 😅

WiktorPieklik commented on src/Utility/TextSplitter.cpp:
```diff
@@ -0,0 +1,33 @@
+                ranges.push_back(std::make_tuple(currenPosition, currenPosition + chains.first - 1));
+                currenPosition = currenPosition + chains.first;
+            }
+        } else {
```
This `else` statement looks weird. I suggest using: 
```cpp
int i = 0;
while(numOfWindows > i) {
//these first if and else statements
++i;
}
```

zawadzki-dawid replied:
Accurate observation!

WiktorPieklik commented on src/Utility/TextSplitter.cpp:
```diff
@@ -0,0 +1,33 @@
+#include "TextSplitter.h"
+#include <iostream>
```
It seems like `TextSplitter` doesn't use `<iostream>`. If I'm right, please remove this `include` statement.

WiktorPieklik commented on src/Utility/TextSplitter.cpp:
```diff
@@ -0,0 +1,33 @@
+    for(int i = 0; i < numOfThreads; i++) {
+        if(windows > i) {
+            if (chains.second >= 0) {
+                ranges.push_back(std::make_tuple(currenPosition, currenPosition + chains.first));
```
Change `push_back` to `emplace_back` here and below.

zawadzki-dawid replied:
Right, `emplace_back` seems to be better pick in this place, thanks for advice :smile:

WiktorPieklik commented on src/Utility/TextSplitter.cpp:
```diff
@@ -0,0 +1,33 @@
+    return std::make_pair(rangeWidth, rangesWithBonus - 1);
+}
+
+std::vector<std::tuple<size_t, size_t>> TextSplitter::splitText(long long int textLength, int patternLength) {
```
Please add some docs telling what user is supposed to get within `TextSplitter::splitText`. We know it for now, but probably we'll forget it in a while.

WiktorPieklik reviewed (changes requested):
Really great job! Your solution works as expected. I left some comments and most of them are connected with code style. Besides that your `TextSplitter` is 🏎️  🎸

szymon-filipiak commented on src/Utility/TextSplitter.cpp:
```diff
@@ -0,0 +1,33 @@
+std::pair<int, int> TextSplitter::lenOfChain(int windows) {
+    int rangeWidth = windows / numOfThreads;
+    int rangesWithBonus = windows % numOfThreads;
+    return std::make_pair(rangeWidth, rangesWithBonus - 1);
```
Should this -1 be here?

szymon-filipiak replied:
I prefer returning a not modified value here and then check `chains.second > 0` not `chains.second >= 0` in line 19.

zawadzki-dawid replied:
Right, I'll change that!

zawadzki-dawid committed:
Remove auto generated comment

zawadzki-dawid committed:
Refactor splitText method to work on std::pair instead on std::tuple

zawadzki-dawid committed:
Replace push_back with emplace_back

zawadzki-dawid committed:
Remove unnecessary iostream include

zawadzki-dawid committed:
Remove one subtraction from return statement in lenOfChain method

zawadzki-dawid committed:
Refactor numOfWindows method to return value of size_t type

zawadzki-dawid committed:
Refactor numOfWindows method to take value of size_t type as first parameter

zawadzki-dawid committed:
Change name of 'windows' variable to 'numOfWindows' to avoid misleading naming

zawadzki-dawid committed:
Change name of 'chains' variable to 'lenOfChain' to avoid misleading naming

zawadzki-dawid committed:
Change name of 'currenPosition' variable to 'currentPosition' in splitText method

zawadzki-dawid committed:
Refactor splitText method to use while loop instead of for loop to make it more neat

zawadzki-dawid committed:
Change lenOfChain method

1. Rename it to lenOfChains
2. Make return type and paramter type to be size_t type

zawadzki-dawid committed:
Add comment which is describing what splitText method does

szymon-filipiak reviewed (approved):
It looks nice. There are still two places where you could use `size_t` not `int` and `long long`, but we can merge this for now.

WiktorPieklik reviewed (approved):
👍
